### PR TITLE
Pooling Stage 2 Interim PR [WiP]

### PR DESCRIPTION
**Pull request by r00ty-tc**
This is the PR (replacing https://github.com/TrinityCore/TrinityCore/pull/13633) for the new pooling framework I've been working on.

Some notes, you will need to populate the two new tables before this will do anything. A demo (as used in the videos I've made) is the following (or course run after the world update SQL has been run to create the tables)

``` SQL
DELETE FROM `pool_template`
WHERE entry = 250000;
INSERT INTO `pool_template`
(entry, min_limit, max_limit, description)
VALUES
(250000,5,16,'Creature Test Fray Island');

DELETE FROM `pool_creature`
WHERE pool_id = 250000;
INSERT INTO `pool_creature_info` 
(pool_id, creature_id, creature_qualifier, chance, spawnMask, phaseMask, modelid, equipment_id, spawntimesecs, corpsetimesecsloot, corpsetimesecsnoloot, spawndist, currentwaypoint, curhealth, curmana, MovementType, npcflag, unit_flags, dynamicflags)
VALUES 
(250000,1501,0,50,1,1,10973,0,0,0,0,0,0,42,0,0,0,0,0),
(250000,1502,0,50,1,1,829,0,0,0,0,0,0,42,0,0,0,0,0);

DELETE FROM `pool_spawns`
WHERE pool_id = 250000;
INSERT INTO `pool_spawns`
(pool_id, guid, map, zoneId, areaId, position_x, position_y, position_z, orientation)
VALUES 
(250000,1,1,0,0,-1661.7,-4368.03,6.84431,2.6558),
(250000,2,1,0,0,-1656.31,-4363.33,6.91614,3.92225),
(250000,3,1,0,0,-1653,-4358.71,6.7165,2.95817),
(250000,4,1,0,0,-1649.65,-4361.44,6.68167,2.34753),
(250000,5,1,0,0,-1648.35,-4355.15,6.52353,3.9517),
(250000,6,1,0,0,-1645.04,-4351.4,6.49343,2.89534),
(250000,7,1,0,0,-1638.05,-4357.79,6.50028,2.85214),
(250000,8,1,0,0,-1641.63,-4347.81,6.55185,0.961291),
(250000,9,1,0,0,-1656.45,-4369.97,6.86095,3.73768),
(250000,10,1,0,0,-1652.62,-4366.14,6.86773,5.59122),
(250000,11,1,0,0,-1645.69,-4365.33,6.75387,3.86923),
(250000,12,1,0,0,-1641.64,-4361.42,6.52736,4.86669),
(250000,13,1,0,0,-1645.06,-4357.74,6.51096,5.44788),
(250000,14,1,0,0,-1641.53,-4353.91,6.48203,2.52031),
(250000,15,1,0,0,-1634.88,-4354.56,6.6078,3.53347),
(250000,16,1,0,0,-1639.17,-4350.37,6.61137,6.2372);
```

It should produce results as per the video https://www.youtube.com/watch?v=UAJA38Fqx1k

You can see you need an entry per creature kind and spawn chance in pool_creature_info and an entry in pool_spawns with the individual spawns and of course the template entry (as per other pools).

Ideally I'd like some testing/bug reports, some performance testing and maybe some analysis where I've gone off track in code.

I know there's some redundant code, I've left it in, in case it's needed for the game object aspect later. I think a proper analysis of redundant code can be performed then. Otherwise, advice and suggestions for improvements are welcome.

Currently working:
- Configuration implemented
- SQL initial table changes
- Load config
- Load databases
- Check chance is total 100 per pool, or all 0.
- Spawns pool according to chance and using spawn points defined (e.g. random mob, random spawn point).
- Dynamic spawn (see the ID in the screenshot)
- Tracks used, not used spawnpoints, functions to locate spawnpoint or creature info per dynamic guid
- Handle despawn
- Handle respawn with new ID, including expedited respawn using config fast timer
- Looting timers (separate for looted, unlooted)
- Loot timer reset (honouring config)
- Optional GUID recycling

To do:
- Handle game objects
- Handle save grid state (optionally)

Later (when stable):
- Integrate with guid from map branch (possibly implement as new PoolMgr as part of map instance)
- Port to 6.x

### CODE REVIEW

**Comment by Shauren** [comment]
This effectively adds "spawn point" functionality

**Comment by pete318** [comment]
Yes, but also dynamic spawns (e.g. the old corpse doesn't go away) allowing for spawn times and corpse decay to be independent.

Plus of course minimum spawns per pool and proper chance of spawn prediction.

**Comment by NorthStrider** [comment]
wouldnt it be possible to use db guids do get the pooling positions ? So the system select the coordinates from the defined database guid.

**Comment by pete318** [comment]
@NorthStrider  Not really no. The creature table holds the information about the creature and the spawn position. However, the dynamic spawn can be a creature info part and any of the spawn parts mixed together.

Hence the change here to check if the creature was spawned by the new pooling manager, if so it asks for the structure info from that manager instead.

**Comment by MrSmite** [comment]
One question and one suggestion:

Q: Does this immediately despawn a corpse after looting? Or will it leave it behind to be skinned if applicable? I didn't see anything in the code specific to looting but I just wanted to bring it up.

Suggestion: Instead of using numbers in function names, use words. It makes it easier to read because in some editors a *1* looks like a lower-case *L*.

eg: *Respawn1Creature()* -> *RespawnOneCreature()*

**Comment by pete318** [comment]
Yep. The naming convention followed the old pool naming. However, I think if moved to map instances then I think only this could be moved (creatures+gos) since daily quest pooling would need to remain as part of the world rather than map.

As for the question. There are two timers. Corse decay when unlooted, and corpse decay once looted.

By default this is set to 5 mins for unlooted, 2 mins once looted. I based this on a very old post on the internet by someone that investigated this. When impleted to 6.x probably newer experiments should be considered to see if the rule still holds.

These timers, have a global setting in the config which applies if the creature has "0" in that time field. Otherwise the creatures can set it at that level.

So in short, corpse decay and respawn are independent and configurable at the creature level.

If you want to test, I have this PR + Some SQL to set up Deathknell (Undead start zone) to pooling, so you can check it out yourself.

Otherwise, if you have no 3.3.5 setup I can make you an account on mine and you can try it out. PM me on irc/forum if you'd like that.

**Comment by Shauren** [comment]
There are more variables to corpse despawning on retail - corpses with epic/rare loot despawn even slower

**Comment by pete318** [comment]
It'd be trivial to add extra parameters for a global value for those (if you see where I implemented it).

Also already it can be changed per creature anyway.

**Comment by MrSmite** [comment]
> Otherwise, if you have no 3.3.5 setup I can make you an account on mine and you can try it out. PM me on irc/forum if you'd like that.

I have a 335 setup, I just haven't launched it in a while. I had to take a break from WoW and Trinity because I was getting burnt out. I still watch issues that I get e-mails about though.

**Comment by click** [comment]
**Note:**
This PR is accepted for inclusion into core, but pending completion of additional code by @Shauren.

**Comment by Naios** [comment]
Any news about this? @Shauren @click @pete318

**Comment by pete318** [comment]
From me, I've had a combination of remote work and illness. But, maybe I'll have some time in the upcoming weekends. There's a fair bit more to be done here really, since there's still some tables referencing by guid and not entry. Plus it really needs to be done with the guid per map code.

The issue there is that I'm developing in 3.3.5 for stability sake. Shauron coded the guid per map into 6.x. Ultimately I'd say both items need to be in both branches anyway though.

**Comment by Naios** [comment]
I still don't get it why you add those 2 new database tables  `pool_creature_info` and `pool_spawns` since most of it is handled in `creature` and `creature_addon already`.

It would be nice if we could talk at irc.rizon.net #trinity about it since your pr looks very interesting.

**Comment by pete318** [comment]
I'll answer here, and can catch me later on IRC for followup.

The reason to use these tables is to split the data. That you define the creature info in pool_creature_info once. Then have an entry multiple times per creature_info entry in pool_spawns. That way, the creature can be chosen at random from the possible creatures, and then the spawn point (that could be shared among different creature types) can be chosen at random.

It would take a lot more overhead to use the existing tables to do this job properly.

On the forum, there's a thread http://community.trinitycore.org/topic/10713-experimental-pooling-branch/

Includes a lot of discussion clarifying the design principles, some videos and the actual design I used to produce it.

**Comment by tkrokli** [comment]
The `pete318:pooling_s2` branch needs updating because of all the changes since 2015-03-05.

**Comment by pete318** [comment]
There's little point updating this yet. The plan isn't to even consider merging this until the map_local_guids becomes part of 3.3.5. At that point, this will need adapting (and of course, incorporating game object changes, and there are issues that need resolving too).

So, this hasn't been forgotten. It's just that there's preparation work required.

**Comment by malcrom** [comment]
Posting this here for you pete. This is a little variation on what you did that would spawn an area in felwood. I moved a few things to spawn template so they wouldn't need to be repeated in spawn pool. Only problem is my idea requires a scripting table for basic creature scripting and path loading. I always wanted that separate from SAI but loadable by SAI for this reason. The ability to load a path by spawn point independent of the combat AI of the creature would make spawning and scripting much easier to do.

The poolIds for each zone would start at zoneId & 000.
The paths for each poolId would start at poolId & 000.
This would make it easier to locate items in db.

```SQL
-- #### Felwood Spawning ####

-- Deadwood Village Spawning template
Set @POOLID := 361000;
DELETE FROM `creature_spawn_template` WHERE entry = @POOLID;
INSERT INTO `creature_spawn_template` (`poolId`, `mapId`, `zoneId`, `areaId`, `phaseId`, `min_limit`, `max_limit`, `respawn`, `description`) VALUES
(@POOLID,1,361,1761,169,10,34,300,'Deadwood Village Spawning');

-- Deadwood Village Spawning pool
DELETE FROM `creature_spawn_pool` WHERE `poolId`=@POOLID;
INSERT INTO `creature_spawn_pool` (`poolId`, `spawnpoint`, `creatureId1`, `creatureId2`, `creatureId3`, `script`, `position_x`, `position_y`, `position_z`, `orientation`, `description`) VALUES
-- No movement
(@POOLID,0,7153,7154,7155,0,3597.07,-1026.9,226.09,3.05,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,1,7153,7154,7155,0,3597.07,-1026.9,226.09,3.05,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,2,7153,7154,7155,0,3622.21,-1130.46,210.79,5.25,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,3,7153,7154,7155,0,3737.23,-1149.34,209.85,3.94,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,4,7153,7154,7155,0,3743.92,-1169.23,209.84,2.46,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,5,7153,7154,7155,0,3618.7,-1188.61,213.86,1.12,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,6,7153,7154,7155,0,3704.07,-1135.35,207.95,5.62,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,7,7153,7154,7155,0,3752.59,-993.49,241.83,3.58,'Deadwood Gardener, Pathfinder, Warrior'),
-- Random movement
(@POOLID,8,7153,7154,7155,1,3533.1,-1108.63,225.34,5.6 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,9,7153,7154,7155,1,3534.61,-1150.68,225.42,5.1 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,10,7153,7154,7155,1,3543.74,-1128.56,222.26,0.27 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,11,7153,7154,7155,1,3609.81,-1164.02,214.18,3.63 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,12,7153,7154,7155,1,3611.9,-1042.81,220.58,1.72 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,13,7153,7154,7155,1,3619.8,-1209.82,217.42,2.36 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,14,7153,7154,7155,1,3636.09,-1168.11,210.07,0.44 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,15,7153,7154,7155,1,3649.05,-1138.72,209.13,3.68 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,16,7153,7154,7155,1,3649.92,-1179.41,211.8,6.28 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,17,7153,7154,7155,1,3653.85,-1028.21,221.65,0.52 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,18,7153,7154,7155,1,3673.6,-1157.89,207.91,2.41 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,19,7153,7154,7155,1,3712.1,-1198.12,207.7,2.06 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,20,7153,7154,7155,1,3716.14,-1015.36,228.82,5.97 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,21,7153,7154,7155,1,3721.86,-1062.5,221.8,3.14 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,22,7153,7154,7155,1,3727.77,-1103.57,213.43,5.04 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,23,7153,7154,7155,1,3773.95,-1196.66,217.44,1.04 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,24,7153,7154,7155,1,3811.52,-1148.61,237.25,6.26 ,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,25,7153,7154,7155,1,3566.68,-1148.24,220.05,0.65,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,26,7153,7154,7155,1,3580.56,-1118.58,219.95,4.01,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,27,7153,7154,7155,1,3583.93,-1076.81,221.25,4.66,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,28,7153,7154,7155,1,3590.91,-1169.25,219.69,3.02,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,29,7153,7154,7155,1,3679.75,-1019.22,224.18,3.54,'Deadwood Gardener, Pathfinder, Warrior'),
-- Pathed movement
(@POOLID,30,7153,7154,7155,2,3577.37,-1073.75,221.76,3.99,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,31,7153,7154,7155,3,3651.72,-1166.22,209,6.22,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,32,7153,7154,7155,4,3719.67,-1133.48,209.19,4.15,'Deadwood Gardener, Pathfinder, Warrior'),
(@POOLID,33,7153,7154,7155,5,3709.65,-1156.61,207.9,4.72 ,'Deadwood Gardener, Pathfinder, Warrior');

-- Scripts for creature_spawn_template 361000 "Yet to work out"

-- Script 1 do random movement
-- Script 2 load path 361000000 repeating
-- Script 3 load path 361000001 repeating
-- Script 4 load path 361000002 repeating
-- Script 5 load path 361000003 repeating

-- Paths for creature_spawn_template 361000
Set @PATH := 361000000;
DELETE FROM `waypoint_data` WHERE `id`=@PATH;
INSERT INTO `waypoint_data` (`id`,`point`,`position_x`,`position_y`,`position_z`,`orientation`,`delay`,`move_type`,`action`,`action_chance`,`wpguid`) VALUES
(@PATH,1,3575.04,-1076.384,221.8051,0,0,0,0,100,0),
(@PATH,2,3567.622,-1098.693,221.5181,0,0,0,0,100,0),
(@PATH,3,3562.087,-1118.363,221.1803,0,0,0,0,100,0),
(@PATH,4,3565.696,-1132.111,221.2504,0,0,0,0,100,0),
(@PATH,5,3576.388,-1143.33,219.6638,0,0,0,0,100,0),
(@PATH,6,3594.952,-1153.905,219.6672,0,0,0,0,100,0),
(@PATH,7,3622.036,-1154.701,211.3212,0,0,0,0,100,0),
(@PATH,8,3647.657,-1169.655,209.7255,0,0,0,0,100,0),
(@PATH,9,3661.205,-1171.2,209.3326,0,0,0,0,100,0),
(@PATH,10,3678.722,-1167.328,208.1529,0,0,0,0,100,0),
(@PATH,11,3702.719,-1169.509,207.4022,0,0,0,0,100,0),
(@PATH,12,3714.411,-1159.28,208.0238,0,0,0,0,100,0),
(@PATH,13,3723.102,-1142.47,209.1148,0,0,0,0,100,0),
(@PATH,14,3731.659,-1124.576,210.9985,0,0,0,0,100,0),
(@PATH,15,3739.66,-1110.396,214.1177,0,0,0,0,100,0),
(@PATH,16,3746.909,-1099.605,217.5009,0,0,0,0,100,0),
(@PATH,17,3744.177,-1084.458,219.0947,0,0,0,0,100,0),
(@PATH,18,3746.693,-1058.91,223.4796,0,0,0,0,100,0),
(@PATH,19,3743.289,-1041.939,225.8803,0,0,0,0,100,0),
(@PATH,20,3725.456,-1022.281,228.6066,0,0,0,0,100,0),
(@PATH,21,3716.919,-1012.96,229.4382,0,0,0,0,100,0),
(@PATH,22,3704.023,-1007.8,228.5495,0,0,0,0,100,0),
(@PATH,23,3690.01,-1010.55,224.4712,0,0,0,0,100,0),
(@PATH,24,3676.192,-1015.828,224.1748,0,0,0,0,100,0),
(@PATH,25,3655.77,-1020.512,222.1121,0,0,0,0,100,0),
(@PATH,26,3640.44,-1027.33,222.6451,0,0,0,0,100,0),
(@PATH,27,3626.348,-1031.823,221.8228,0,0,0,0,100,0),
(@PATH,28,3612.275,-1040.839,221.0037,0,0,0,0,100,0),
(@PATH,29,3599.829,-1051.741,219.7981,0,0,0,0,100,0),
(@PATH,30,3587.512,-1062.276,220.4841,0,0,0,0,100,0);

Set @PATH := 361000001;
DELETE FROM `waypoint_data` WHERE `id`=@PATH;
INSERT INTO `waypoint_data` (`id`,`point`,`position_x`,`position_y`,`position_z`,`orientation`,`delay`,`move_type`,`action`,`action_chance`,`wpguid`) VALUES
(@PATH,1,3651.716,-1166.215,208.9989,0,0,0,0,100,0),
(@PATH,2,3641.558,-1165.332,209.4411,0,0,0,0,100,0),
(@PATH,3,3633.445,-1167.879,210.3903,0,0,0,0,100,0),
(@PATH,4,3627.21,-1177.31,211.7565,0,0,0,0,100,0),
(@PATH,5,3631.898,-1170.092,210.6403,0,0,0,0,100,0),
(@PATH,6,3641.504,-1165.535,209.4476,0,0,0,0,100,0);

Set @PATH := 361000002;
DELETE FROM `waypoint_data` WHERE `id`=@PATH;
INSERT INTO `waypoint_data` (`id`,`point`,`position_x`,`position_y`,`position_z`,`orientation`,`delay`,`move_type`,`action`,`action_chance`,`wpguid`) VALUES
(@PATH,1,3697.57,-1151.694,207.032,0,0,0,0,100,0),
(@PATH,2,3676.109,-1155.196,208.0412,0,0,0,0,100,0),
(@PATH,3,3659.861,-1154.675,208.9906,0,0,0,0,100,0),
(@PATH,4,3644.097,-1148.313,209.9757,0,0,0,0,100,0),
(@PATH,5,3627.295,-1150.123,211.3212,0,0,0,0,100,0),
(@PATH,6,3606.05,-1143.36,213.6902,0,0,0,0,100,0),
(@PATH,7,3595.56,-1131.1,217.1226,0,0,0,0,100,0),
(@PATH,8,3587.919,-1119.175,218.7625,0,0,0,0,100,0),
(@PATH,9,3586.05,-1105.474,220.105,0,0,0,0,100,0),
(@PATH,10,3588.586,-1096.266,220.4756,0,0,0,0,100,0),
(@PATH,11,3590.767,-1085.894,221.7661,0,0,0,0,100,0),
(@PATH,12,3598.574,-1075.09,219.8293,0,0,0,0,100,0),
(@PATH,13,3608.946,-1062.424,219.4926,0,0,0,0,100,0),
(@PATH,14,3615.667,-1054.927,219.1191,0,0,0,0,100,0),
(@PATH,15,3629.311,-1046.101,220.8212,0,0,0,0,100,0),
(@PATH,16,3645.503,-1041.352,221.3114,0,0,0,0,100,0),
(@PATH,17,3657.924,-1036.394,221.1315,0,0,0,0,100,0),
(@PATH,18,3670.764,-1029.754,222.6582,0,0,0,0,100,0),
(@PATH,19,3691.562,-1026.028,223.2998,0,0,0,0,100,0),
(@PATH,20,3705.705,-1031.286,223.8864,0,0,0,0,100,0),
(@PATH,21,3717.623,-1051.01,222.7746,0,0,0,0,100,0),
(@PATH,22,3726.91,-1063.109,221.2639,0,0,0,0,100,0),
(@PATH,23,3735.42,-1079.583,218.7634,0,0,0,0,100,0),
(@PATH,24,3730.652,-1090.352,216.965,0,0,0,0,100,0),
(@PATH,25,3723.369,-1104.703,213.0214,0,0,0,0,100,0),
(@PATH,26,3722.022,-1118.559,210.1962,0,0,0,0,100,0),
(@PATH,27,3720.389,-1132.356,209.2753,0,0,0,0,100,0),
(@PATH,28,3712.491,-1144.754,208.3352,0,0,0,0,100,0);

Set @PATH := 361000003;
DELETE FROM `waypoint_data` WHERE `id`=@PATH;
INSERT INTO `waypoint_data` (`id`,`point`,`position_x`,`position_y`,`position_z`,`orientation`,`delay`,`move_type`,`action`,`action_chance`,`wpguid`) VALUES
(@PATH,1,3709.52,-1140.89,207.9481,0,0,0,0,100,0),
(@PATH,2,3709.652,-1156.609,207.8988,0,0,0,0,100,0);
```

**Comment by ShinDarth** [comment]
Are there any news about this?

**Comment by jackpoz** [comment]
Re-open this PR when you plan to go on working on it, there's little point leaving it here getting dust
